Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,519 @@
+      remoteAddress: context.remoteAddress,
+      userInfoRef: userInfoRef,
+      interceptors: interceptors,
+      onRequestPart: self.receiveInterceptedPart(_:),
+      onResponsePart: self.sendInterceptedPart(_:promise:)
+    )
+  }
+
+  // MARK: - Public API: gRPC to Handler
+
+  @inlinable
+  public func receiveMetadata(_ headers: HPACKHeaders) {
+    self.interceptors.receive(.metadata(headers))
+  }
+
+  @inlinable
+  public func receiveMessage(_ bytes: ByteBuffer) {
+    do {
+      let message = try self.deserializer.deserialize(byteBuffer: bytes)
+      self.interceptors.receive(.message(message))
+    } catch {
+      self.handleError(error)
+    }
+  }
+
+  @inlinable
+  public func receiveEnd() {
+    self.interceptors.receive(.end)
+  }
+
+  @inlinable
+  public func receiveError(_ error: Error) {
+    self.handleError(error)
+    self.finish()
+  }
+
+  @inlinable
+  public func finish() {
+    switch self.state {
+    case .idle:
+      self.interceptors = nil
+      self.state = .completed
+
+    case let .observing(_, context):
+      context.statusPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+    case .completed:
+      self.interceptors = nil
+    }
+  }
+
+  // MARK: - Interceptors to User Function
+
+  @inlinable
+  internal func receiveInterceptedPart(_ part: GRPCServerRequestPart<Request>) {
+    switch part {
+    case let .metadata(headers):
+      self.receiveInterceptedMetadata(headers)
+    case let .message(message):
+      self.receiveInterceptedMessage(message)
+    case .end:
+      self.receiveInterceptedEnd()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      // Make a context to invoke the user handler with.
+      //
+      // The user function accepts a protocol-type of `AsyncServerCallContext`.
+      // For now we use `_StreamingResponseCallContext` but we will move to a dedicated concrete type soon.
+      //
+      // TODO: Update to use dedicated `AsyncServerCallContext` type.
+      let context = _StreamingResponseCallContext<Request, Response>(
+        eventLoop: self.context.eventLoop,
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef,
+        compressionIsEnabled: self.context.encoding.isEnabled,
+        closeFuture: self.context.closeFuture,
+        sendResponse: self.interceptResponse(_:metadata:promise:)
+      )
+
+      // Create a request stream to pass to the user function and capture the
+      // handler in the updated state to allow us to produce more results.
+      let requestStream = GRPCAsyncStream<Request>(AsyncThrowingStream { continuation in
+        self.state = .observing({ streamEvent in
+          switch streamEvent {
+          case let .message(request): continuation.yield(request)
+          case .end: continuation.finish()
+          }
+        }, context)
+      })
```

Comment:
You're right that this type is a bit obscure because it is the glue layer from NIO world to async-await world.

This handler uses an EL internally because it interacts with the rest of the internal GRPC library. Its internal implementation is similar in structure to the extant `BidirectionalStreamingServerHandler`.

In order to call the user function, we construct an `AsyncSequence` of requests and set the `state` to `.observing` with a closure that handles incoming stream events (message or end).

The closure passed into the `.obeserving` enum value will always be called on the right EL.

If I've misunderstood this then I guess we'll need to do something different here. 

Worth noting that when we have #1245 we will not be doing things this way and the code here is a stopgap so may not be worth agonising over.